pkg/targets/targetspb/rpc.proto

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+syntax = "proto3";
+package thanos;
+
+import "store/storepb/types.proto";
+import "gogoproto/gogo.proto";
+import "store/labelpb/types.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "targetspb";
+
+option (gogoproto.sizer_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_getters_all) = false;
+
+// Do not generate XXX fields to reduce memory footprint and opening a door
+// for zero-copy casts to/from prometheus data types.
+option (gogoproto.goproto_unkeyed_all) = false;
+option (gogoproto.goproto_unrecognized_all) = false;
+option (gogoproto.goproto_sizecache_all) = false;
+
+/// Targets represents API that is responsible for gathering targets and their states.
+service Targets {
+    /// Targets has info for all targets.
+    /// Returned targets are expected to include external labels.
+    rpc Targets (TargetsRequest) returns (stream TargetsResponse);
+}
+
+message TargetsRequest {
+    enum State {
+        ANY = 0;
+        /// This will make sure strings.ToLower(.String()) will match 'active' and 'dropped' values for
+        /// Prometheus HTTP API.
+        ACTIVE = 1;
+        DROPPED = 2;
+    }
+    State state = 1;
+    PartialResponseStrategy partial_response_strategy = 2;
+}
+
+message TargetsResponse {
```

Comment:
Can we ensure that targets are sorted by something which will allow us to merge them on global level without fully buffer ing all frames? (: If that's already the case can we put comment saying about sorting API invariant? (:

Reply:
I think this is already being done when deduplicating targets but it's worth adding a comment here nonetheless (and maybe even a test).

Reply:
I'm absolutely out of understanding the idea.

Probably because it connected with "implementing streaming" which I'm can't understand either. :-D